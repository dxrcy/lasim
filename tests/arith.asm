.ORIG x3000

    and r1, r1, x0
    REG

    add r1, r1, x1
    REG

    ld r0, Hex0010
    add r1, r1, r0
    REG

    ld r0, Dec10
    add r1, r1, r0
    REG

    and r1, r1, x1
    REG
    and r1, r1, x1
    REG

    not r1, r1
    REG

    and r1, r1, x0
    REG
    and r1, r1, x0
    REG

    ld r0, Hexd69b
    add r1, r1, r0
    REG

    not r1, r1
    REG
    not r1, r1
    REG

    ld r0, Hex2d5a
    and r1, r1, r0
    REG

    HALT

Dec10   .FILL #10
Hex0010 .FILL x10
Hexd69b .FILL xd69b
Hex2d5a .FILL x2d5a

.END
