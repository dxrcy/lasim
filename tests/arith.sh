#!/bin/sh

tests_dir="$(dirname $0)"
out_dir="$tests_dir/out"

lasim() {
    "$tests_dir/../lasim" $@ || exit $?;
}
alias diff='delta'

asm_file="$tests_dir/arith.asm"
obj_file="$out_dir/arith.obj"
output_actual_file="$out_dir/arith.actual"
output_expected_file="$tests_dir/arith.expected"

extract_reg() {
    sed -n 's/ *. *r1 *\([^ ]*\).*/\1/p'
}

lasim -a "$asm_file" -o "$obj_file" || exit $?
./lasim -x "$obj_file" | extract_reg > "$output_actual_file"

diff "$output_expected_file" "$output_actual_file"
if [ $? -eq 0 ];
    then printf '\x1b[32mpass\x1b[0m\n'
    else printf '\x1b[31mFAIL\x1b[0m\n'
fi

