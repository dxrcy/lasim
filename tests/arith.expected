0x0000
0x0001
0x0011
0x001b
0x0001
0x0001
0xfffe
0x0000
0x0000
0xd69b
0x2964
0xd69b
0x041a
